olp-cpp-sdk-dataservice-read/src/repositories/CatalogRepository.cpp

```diff
@@ -238,6 +239,90 @@ client::CancellationToken CatalogRepository::getLatestCatalogVersion(
       [cancel_context]() { cancel_context->CancelOperation(); });
 }
 
+MetadataApi::CatalogVersionResponse CatalogRepository::GetLatestVersion(
+    const client::HRN& catalog,
+    client::CancellationContext cancellation_context,
+    const DataRequest& data_request, client::OlpClientSettings settings) {
+  using namespace client;
+
+  repository::CatalogCacheRepository repository(catalog, settings.cache);
+
+  auto fetch_option = data_request.GetFetchOption();
+  if (fetch_option != OnlineOnly) {
+    auto cached_version = repository.GetVersion();
+    if (cached_version) {
+      OLP_SDK_LOG_INFO_F(kLogTag, "cache catalog '%s' found!",
+                         data_request.CreateKey().c_str());
+      return cached_version.value();
+    } else if (fetch_option == CacheOnly) {
+      OLP_SDK_LOG_INFO_F(kLogTag, "cache catalog '%s' not found!",
+                         data_request.CreateKey().c_str());
+      return ApiError(
+          ErrorCode::NotFound,
+          "Cache only resource not found in cache (catalog version).");
+    }
+  }
+
+  const auto timeout = settings.retry_settings.timeout;
+
+  auto metadata_api =
+      ApiClientLookup::LookupApi(catalog, cancellation_context, "metadata",
+                                 "v1", fetch_option, std::move(settings));
+
+  if (!metadata_api.IsSuccessful()) {
+    return metadata_api.GetError();
+  }
+
+  const client::OlpClient& client = metadata_api.GetResult();
+
+  Condition condition{std::chrono::seconds{timeout}};
+
+  // when the network operation took too much time we cancel it and exit
+  // execution, to make sure that network callback will not access dangling
+  // references we protect them with atomic bool flag.
+  auto interest_flag = std::make_shared<std::atomic_bool>(true);
+
+  MetadataApi::CatalogVersionResponse version_response;
+  cancellation_context.ExecuteOrCancelled([&]() {
+    auto token = MetadataApi::GetLatestCatalogVersion(
+        client, -1, data_request.GetBillingTag(),
+        [&, interest_flag](MetadataApi::CatalogVersionResponse response) {
+          if (interest_flag->exchange(false)) {
+            version_response = std::move(response);
+            condition.Notify();
+          }
+        });
+
+    return client::CancellationToken([&, interest_flag, token]() {
+      if (interest_flag->exchange(false)) {
+        token.cancel();
+        condition.Notify();
+      }
+    });
+  });
+
+  if (!condition.Wait()) {
```

Comment:
If context was cancelled before this block, the execute will not be triggered and the wait condition will fail eventually (but we rather want to return 'cancelled' state)
Consider to add "condition.Notify();" as the cancel callback of ExecuteOrCancelled

Reply:
Thanks! 
Added test case for that, it is broken at the moment, but eventually, after https://github.com/heremaps/here-olp-sdk-cpp/pull/244 merged, it must be ok.